### ISSUE TITLE
Import keyboard layout from desktop environment 

### ISSUE THREAD

**Issue by magicalraccoon** [Plagman/gamescope#203]
I'm normally a dvorak user, but when I use Gamescope the layout remains qwerty. 

The keyboard mapping and layout should be imported from the current de/wm to create a seamless experience. Is this something the project could handle or will I need to get used to gaming with qwerty?

**Comment by emersion** [comment]
This is definitely something gamescope should do. We need to set the correct layout on the `wlr_keyboard`.

**Comment by emersion** [comment]
Workaround: `export XKB_DEFAULT_LAYOUT=<layout>` before starting gamescope (see `man xkeyboard-config` for a list of layout names).

**Comment by IDeathByte** [comment]
not sure about create new issue, then i post here

as i understand, by default gamescope doesn't import layout from main session
this take us problem with layout switch (en\ru) it didn't changes by switch-key (capslock on my Manjaro KDE config, wayland session if it important)

**Comment by itspngu** [comment]
> Workaround: `export XKB_DEFAULT_LAYOUT=<layout>` before starting gamescope (see `man xkeyboard-config` for a list of layout names).

I can confirm that this works for gamescope itself, but when using it to run Steam's "Big Picture" mode the latter crashes, which might be a common source of confusion given gamescope's heritage. I can not type "@" in the Big Picture login mask (on German keyboards "@" is produced by holding AltGr and pressing Q) regardless of wether gamescope is running or not.

https://github.com/ValveSoftware/steam-for-linux/issues/5188

**Comment by Figuera** [comment]
I had some success by using `XKB_DEFAULT_LAYOUT` but that is not enough I also need to set the Layout Variant and setting `XKB_DEFAULT_VARIANT` do not seems to help.

It might be my bias, but it seems to me that the compositor would benefit greatly from a configuration file. These are stuff that should be easily managed there. I would also like to be able to customize the key bindings. Alt+F4 is fine but I use something else in my Desktop Environment and would be nice to be able to change it somewhere.

**Comment by class101** [comment]
I just broke the boot of my Steam Deck :(

Tested the following lines in `/etc/environment`

```
XKBLAYOUT=fr
XKBVARIANT=fr
XKB_DEFAULT_LAYOUT=fr
XKB_DEFAULT_VARIANT=fr
```

Luckily I have been able to boot from an old save, edit revert, and it boots again

I was initially trying to find a solution on why the "fr" layout I set in Desktop mode is not preserved in gaming mode

But is respected in gaming mode IN a open KDE session

I can get my MX Key Mini bluetooth keyboard fr in desktop , fr in gaming mode, it is alwyas us in gaming mode :(

I'd expected to fix it with the variables, but its crashes at boot

**Comment by class101** [comment]
Ok my mistake was to setup 

```
XKB_DEFAULT_LAYOUT=fr
XKB_DEFAULT_VARIANT=fr
```

instead of 

```
XKB_DEFAULT_LAYOUT=fr
XKB_DEFAULT_VARIANT=azerty
```

This resulted in the following gamescope crash

```
wlserver: Running compositor on wayland display gamescope-0
wlserver: [backend/headless/backend.c:18] Starting headless backend
xkbcommon: ERROR: Couldnt process include statement for fr(fr)
xkbcommon: ERROR: Abandoning symbols file (unnamed)
xkbcommon: ERROR: Failed to compile xkb_symbols
xkbcommon: ERROR: Failed to compile keymap
Segmentation fault (core dumped)
```

I have the feeling nobody is concerned by the fact, defining an environment variable could result in a gamescope crash and so on, a Steam Deck failure to boot demonstration

**Comment by emersion** [comment]
Exiting with an error would be the expected behavior.

**Comment by class101** [comment]
There is no possibility to default to US and log a warning ?

**Comment by Glog78** [comment]
I tried on ADOM:
XKB_DEFAULT_LAYOUT=de XKBLAYOUT=de gamescope -Y -e -W 1824 -H 1026 -r 60 -w 1920 -h 1080 -- %command%

still the keyboard is qwerty instead of qwertz ...

**Comment by class101** [comment]
> I tried on ADOM:
> 
> XKB_DEFAULT_LAYOUT=de XKBLAYOUT=de gamescope -Y -e -W 1824 -H 1026 -r 60 -w 1920 -h 1080 -- %command%
> 
> 
> 
> still the keyboard is qwerty instead of qwertz ...

I think you forgot to set `XKB_DEFAULT_VARIANT=dsb_qwertz` see here for more details https://github.com/ValveSoftware/SteamOS/issues/798

**Comment by Glog78** [comment]
> > I tried on ADOM:
> > XKB_DEFAULT_LAYOUT=de XKBLAYOUT=de gamescope -Y -e -W 1824 -H 1026 -r 60 -w 1920 -h 1080 -- %command%
> > still the keyboard is qwerty instead of qwertz ...
> 
> I think you forgot to set `XKB_DEFAULT_VARIANT=dsb_qwertz` see here for more details [ValveSoftware/SteamOS#798](https://github.com/ValveSoftware/SteamOS/issues/798)

Even my solution worked once i switched the steam language to german too ...

**Comment by VladislavNekto** [comment]
But, hey, what if i type in something like Cyrillic, that means, i need possibility to change layout, not just set russian when start

**Comment by IDdozer** [comment]
> But, hey, what if i type in something like Cyrillic, that means, i need possibility to change layout, not just set russian when start

My launch parameters: `XKB_DEFAULT_LAYOUT=ru,us XKB_DEFAULT_OPTIONS=grp:lctrl_lshift_toggle gamemoderun gamescope -e -f -H 1080 -O card0-DVI-D-1 -- %command%`

`left ctrl + left shift` switches language

**Comment by Glog78** [comment]
> > > > I tried on ADOM:
> > > > XKB_DEFAULT_LAYOUT=de XKBLAYOUT=de gamescope -Y -e -W 1824 -H 1026 -r 60 -w 1920 -h 1080 -- %command%
> > > > still the keyboard is qwerty instead of qwertz ...
> > > 
> > > 
> > > I think you forgot to set `XKB_DEFAULT_VARIANT=dsb_qwertz` see here for more details [ValveSoftware/SteamOS#798](https://github.com/ValveSoftware/SteamOS/issues/798)
> > 
> > 
> > Even my solution worked once i switched the steam language to german too ...
> 
> have you had any luck without changing the steam language? i've set up
> 
> ```
> XKB_DEFAULT_LAYOUT=de
> # initially without a variant, also tested with
> XKB_DEFAULT_VARIANT=deadgraveacute
> ```
> 
> in `/etc/environment` but gamescope still is using qwerty for me :(

i didn't needed to switch back. i don't mind my client to be in german too but yeah it's only half of a solution

**Comment by pvdrz** [comment]
:wave: I have a related issue with the `latam` keyboard layout. What happens is that gamescope doesn't recognize the `<` key at all. This key is called '<LSGT>' and it is assigned the `94` code. All the other keys and mappings of the layout work correctly when setting `XKB_DEFAULT_LAYOUT=latam` but for some odd reason this single key doesn't work. I tried setting `XKB_DEFAULT_MODEL` to both `pc104` and `pc105` without success either. 

Has anyone experienced something similar or has any idea on how to debug this?

**Comment by okfruit** [comment]
> What happens is that gamescope doesn't recognize the `<` key at all

Same here

**Comment by Chris2000SP** [comment]
I had that issue with a Steam Linux native game "X4 Foundations" using gamescope under wayland. That workaround with "XKB_DEFAULT_LAYOUT=de" in the Steam Launch option for X4 did work. Thanks for that. I dunno if that matters on Proton games too?

**Comment by murlakatamenka** [comment]
@Chris2000SP so you say if you launch Steam with XKB_DEFAULT_LAYOUT set, `gamescope` does properly inherit it and works as it should keyboard shortcuts wise, am I correct?

**Comment by Chris2000SP** [comment]
If i do not set this, use wayland and gamescope, i get "EN" querty layout in game. Here is what i set in X4 Foundations Launch Options:
`AMD_VULKAN_ICD=RADV XKB_DEFAULT_LAYOUT=de /usr/bin/gamescope -f -W 3840 -H 2160 -- /usr/bin/mangohud %command%`
With "XKB_DEFAULT_LAYOUT=de" fixes the wrong keyboard layout.
I have Arch Linux and KDE by the way.

**Comment by brndd** [comment]
> 👋 I have a related issue with the `latam` keyboard layout. What happens is that gamescope doesn't recognize the `<` key at all. This key is called `<LSGT>` and it is assigned the `94` code. All the other keys and mappings of the layout work correctly when setting `XKB_DEFAULT_LAYOUT=latam` but for some odd reason this single key doesn't work. I tried setting `XKB_DEFAULT_MODEL` to both `pc104` and `pc105` without success either.
> 
> Has anyone experienced something similar or has any idea on how to debug this?

This problem should be fixed after this commit makes its way downstream to you: https://github.com/ValveSoftware/gamescope/pull/1009

**Comment by Gpinchon** [comment]
On OpenSUSE adding a export_kb_layout.sh to /etc/profile.d/ with this content fixes the issue
```
export XKB_DEFAULT_LAYOUT=fr
export XKB_DEFAULT_VARIANT=azerty
```

**Comment by bljordan** [comment]
I don't have a solution, but just adding another report of the issue.

I am in hyprland and trying the solutions above to no avail.

I am trying to set `kb_options = caps:numlock`, `XKB_DEFAULT_OPTIONS=caps:numlock`, or `XKBOPTIONS=caps:numlock`.

**Comment by class101** [comment]
> I don't have a solution, but just adding another report of the issue.
> 
> 
> 
> I am in hyprland and trying the solutions above to no avail.
> 
> 
> 
> I am trying to set the equivalent of `kb_options = caps:numlock` by setting these env vars: `XKB_DEFAULT_OPTIONS=caps:numlock` or `XKBOPTIONS=caps:numlock`.
> 
> 

Few things to try, use double quotes to prevent : from being interpreted 

Else

Try to add in `/etc/X11/xorg.conf.d/00-keyboard.conf`

`setxkbmap -option "caps:numlock"`

To note that based on ArchWiki, setxkbmap is not persistent for external keyboards except if you specify the keyboard device Id, so it is possible that you can do it with setkbdmap manually without a xinit like execution

**Comment by Gpinchon** [comment]
> I am in hyprland and trying the solutions above to no avail.

Yeah, it seems that it doesn't work anymore with the version `3.13.3-1.1`, plus it's very buggy, The Sims 3 appears squished now... Looking as to how to restore it.

**Comment by Martan404** [comment]
This is a big problem on the Steam Deck. When connecting a keyboard to the in Gamemode it will use US qwerty layout with no way of changing it in Gamemode. Works fine in Desktop mode

**Comment by GVKAWESOME** [comment]
using `XKB_DEFAULT_LAYOUT=dk gamescope` in the launch options of Counter-Strike 2 work with danish layout and æøå

**Comment by Dillongc21** [comment]
Anyone have a workaround to this in Wayland? I'm currently running in Gnome Wayland and it's very annoying to not be able to use my Caps key as Escape. Gamescope doesn't appear to respect XKB environment variables above or gsettings options when running in Wayland.

**Comment by murlakatamenka** [comment]
@Dillongc21 you can use a universal solution like https://github.com/xremap/xremap that utilizes `evdev`/`uinput` thus is distro/session agnostic.

**Comment by Dillongc21** [comment]
> [@Dillongc21](https://github.com/Dillongc21) you can use a universal solution like https://github.com/xremap/xremap that utilizes `evdev`/`uinput` thus is distro/session agnostic.

Just set up xremap and it worked like a charm. Thanks!!!

**Comment by esrevinue** [comment]
Having this issue in sway. Is there a nice command line option I can pass that just makes gamescope inherit sway's bindings?

**Comment by Patafix** [comment]
Adding this env variables work for me on hyprland + fedora 42

```
  "XKB_DEFAULT_LAYOUT=fr"
  "XKB_DEFAULT_VARIANT=azerty"
```

**Comment by awsms** [comment]
> Adding this env variables work for me on hyprland + fedora 42
> 
> ```
>   "XKB_DEFAULT_LAYOUT=fr"
>   "XKB_DEFAULT_VARIANT=azerty"
> ```

Does it always respect the layout if you quit/relaunch several times a game on Steam?

**Comment by PeeK1e** [comment]
I managed to set the layout with this:
```bash
env XKB_DEFAULT_VARIANT=nodeadkeys XKB_DEFAULT_MODEL=pc105 XKB_DEFAULT_LAYOUT=de LD_PRELOAD="" gamemoderun gamescope -H1440 -W2560 -r 180 --adaptive-sync --force-grab-cursor -- %command%
```
while the only important part should be `env XKB_DEFAULT_VARIANT=nodeadkeys XKB_DEFAULT_MODEL=pc105 XKB_DEFAULT_LAYOUT=de gamescope -- %command%`

**Comment by awsms** [comment]
Just realized something, the games do recognize my AZERTY keyboard as a QWERTY keyboard in the settings keybinds/in-game keybinds (if I bind `Z` to an action, it recognizes it as `W`, if I type `Z` in-game, it triggers the `W` action).
However, when there is a text area in-game (like an option to report a bug, or text chat), it uses the AZERTY layout (if I type `Z`, it inputs a `Z` in the text box).
Someone has a clue what might be happening?

**Comment by mroopth** [comment]
Here's what I use as a launch option in steam for my dvorak board:

`env XKB_DEFAULT_LAYOUT=us XKB_DEFAULT_MODEL=pc105 XKB_DEFAULT_VARIANT=dvorak LD_PRELOAD="" gamescope -W 2560 -H 1440 -f --force-grab-cursor --adaptive-sync -- %command%`

Not sure if `XKB_DEFAULT_MODEL=pc105` is necessary, but it doesn't make the game crash so ¯\\\_(ツ)_/¯

**Comment by kevinholtkamp** [comment]
In addition to the keyboard layout itself, one issue I also have is that I have switched my ESC and Caps Lock keys.

In gamescope using `env XKB_DEFAULT_VARIANT=nodeadkeys XKB_DEFAULT_MODEL=pc105 XKB_DEFAULT_LAYOUT=de LD_PRELOAD="" `

When in game, pressing ESC works as ESC in the game, but activates Caps Lock in my OS. Typing in game still works normal, while the indicator light is turned on, and in my OS Caps Lock is on

**Comment by ninetailedtori** [comment]
> Replying to https://github.com/ValveSoftware/gamescope/issues/203#issuecomment-3967263806

One thing possible is to remap it not in the user space side but to remap the firmware to the keyboard's switch matrix itself, which works flawlessly, as there's no disparity.

**Comment by kevinholtkamp** [comment]
> Replying to https://github.com/ValveSoftware/gamescope/issues/203#issuecomment-3976579948

That sounds like it would work, but I have no idea how to do it, and from my perspective it seems a little overkill compared to gamescope just passing the keyboard through directly

**Comment by Sarisan** [comment]
Any workaround for people who speak multiple languages? Setting `XKBLAYOUT=us,ru` and `XKBOPTIONS=grp:alt_shift_toggle` doesn't seem to work at all.

**Comment by winter-sunny-morning** [comment]
> Есть ли какое-нибудь решение для людей, говорящих на нескольких языках? Настройки `XKBLAYOUT=us,ru` и `XKBOPTIONS=grp:alt_shift_toggle`Похоже, это вообще не работает.

This configuration should work:
`HOST_LC_ALL=ru_RU.UTF-8 XKB_DEFAULT_LAYOUT=us,ru XKB_DEFAULT_OPTIONS=grp:alt_shift_toggle`
At least it works for me in Genshin Impact

**Comment by wismill** [comment]
Hi, dev of [libxkbcommon](https://xkbcommon.org/) (the library that handle keyboard layouts) here. While I think end users should not have to care about setting the layout env variables manually[^1], let me provide some guidance.

**Disclaimer:** I am not much familiar with gamescope and discovered this issue only by chance.

- The correct environment variables *names* are `XKB_DEFAULT_*` (note the underscores), see the [gamescope implementation](https://github.com/ValveSoftware/gamescope/blob/f8b33d38c5acc35825c7966b208222770c4a623e/src/wlserver.cpp#L1973-L1977).
- If you want the same setup than your normal environment, the environment variables *values* are located differently depending on your DE: e.g. for KDE Plasma, it is `~/.config/kxkbrc`.
- To list all the layout/options available on you computer, use the command `xkbcli list --load-exotic` (requires the package `libxkbcommon-tools` or similar).
- There is also the more visual [layout gallery](https://xkeyboard-config.freedesktop.org/layouts/gallery) for the standard layouts.

[^1]: In my opinion gamescope should set the same layout config than the session *by default*, with possible customization via en variables or whatever.

**Comment by ninetailedtori** [comment]
Have we got a PR for this yet? If not, I'll toss one up.

**Comment by NicMi** [comment]
I want to ask if someone could make work the Altgr key for latam variation, if so could you share about it?

**Comment by ninetailedtori** [comment]
> ## Footnotes
> 
> 1. In my opinion gamescope should set the same layout config than the session _by default_, with possible customization via en variables or whatever.

I'll push a fix for this, establishing a context from the base session then only using the vars to overwrite is primarily the solution, and it's quite doable, so it should be a quick PR!

**Comment by llbranco** [comment]
this solved my keyboard layout (Brazilian ABNT2) but not my numlock state

env XKB_DEFAULT_LAYOUT=br LD_PRELOAD="" gamescope -f --force-grab-cursor -w 1920 -h 1080 -W 1920 -H 1080 -- env PROTON_USE_NTSYNC=1 %command%

**Comment by victorz** [comment]
All these solutions that use environment variables fail to solve the case where I want to switch layout *while playing*. When I type to my friends, I want to use the layout of my native language. When I type in English to random people online/in-game, I want to use an English layout.

Environment variables don't fix this use case as far as I can tell, since that would just set the layout statically for the lifetime of the gamescope process.

Please correct me if I'm wrong. 🙏️

**Comment by ninetailedtori** [comment]
Theoretically I handle that in mine. There's a listener for host keymap changes, and that should update the client keymap. The variables load in only at init, so they build the initial keymap if they are set, otherwise we grab the host key map, otherwise we build the default.

1. Init
- Are vars defined, if so generate local variables and force rebuild flag.

- if rebuild flag set, fetch all, or fetch default vars. Build keymap
- if unset, grab keymap from host
- if returned object is empty, build default as fallback.
- assert and fail if this final keymap object is empty.

- set keymap to whatever is in the final keymap object.

- register a listener for keymap changes with the listeners for the input and output changes as well.

Listener: if keymap changes, grab it from the hosting thread. If it's Wayland, we support returning this already, the stuff I did, returns it from the Wayland thread object, else return nothing (unimplemented for other backends)

**Comment by wismill** [comment]
>  When I type to my friends, I want to use the layout of my native language. When I type in English to random people online/in-game, I want to use an English layout.

@victorz @ninetailedtori There are 2 concepts here:
<dl>
<dt>Layout</dt>
<dd>

- A mapping of keycodes and shift levels to symbols/actions.
- It is one of the components of a keymap.
</dd>
<dt>Keymap</dt>
<dd>

The *complete* definition of the levels, mapping of raw keycodes to symbols/actions, etc. It fully defines the behavior of a keyboard.

A keymap have one or multiple layouts, identified by their indices.

Exactly one layout layout is active at all times.
</dd>
</dl>

If I understand @ninetailedtori correctly, @victorz just has to configure their host with the 2 layouts they need (native + English), then they need no environment variables and have the same keyboard configuration in gamescope. Switching layouts should be possible with the same key combos if enabled on host, otherwise whatever layout switcher the DE proposes.

No keymap reset is required, since the keymap already contain both required layouts. However, gamescope will listen to keyboard state changes (modifiers, layout index) to sync its active layout with host.

**Comment by ninetailedtori** [comment]
Ah yes, that's a much more correct way of putting it! Yes, we sync up the active host layout on layout change events, to the gamescope session as long as the host keymap was not overridden at the init build-time. 
If you wanted a more gamescope-native solution, we'd need to implement our own client keymap wrapper class, layout handlers, then hook them up to keybinding listeners to control it. 
This is possible with a refactor to the current gamescope input model, but out of scope for this current ticket, really. Future me issues...

**Comment by wismill** [comment]
@victorz Note that the environment variables enables to configure multiple layouts using a comma-separated list, e.g. `XKB_DEFAULT_LAYOUT=xx,us`. If you specify a variant, be careful to have match the layout list and leave a blank if no variant is defined, e.g.: 

```
XKB_DEFAULT_LAYOUT="xx,us" XKB_DEFAULT_VARIANT="yyy,"
```

Note the empty variant for the `us` layout: nothing after the last comma.

**Comment by victorz** [comment]
> If I understand [@ninetailedtori](https://github.com/ninetailedtori) correctly, [@victorz](https://github.com/victorz) just has to configure their host with the 2 layouts they need (native + English), then they need no environment variables and have the same keyboard configuration in gamescope. Switching layouts should be possible with the same key combos if enabled on host, otherwise whatever layout switcher the DE proposes.
> 
> No keymap reset is required, since the keymap already contain both required layouts. However, gamescope will listen to keyboard state changes (modifiers, layout index) to sync its active layout with host.

Well this isn't working at all right now for me. I've been testing running Steam games using gamescope, and:

1. the games are not made aware of my switching keyboard layout language
2. I am not at all able to type whatsoever, into any window in the Steam overlay, even though Steam itself is not running using gamescope but only the games themselves. This might be logical somehow, but it's definitely not desirable, of course.

So something is definitely not working as intended here.

I'm running the Niri compositor, with AMDGPU driver on Arch Linux, and have `xwayland-satellite` installed.

Let me know if I'm still misunderstanding something! 😅️

**Comment by MathiasLui** [comment]
I'm on Fedora running CS2 inside of gamescope to do stretching easily - my only keyboard layout on my host system is qwertz (german) but the layout in-game is qwerty.

It should indeed inherit the host layout by default instead
